ci/cd/utils/artifact_repository.py

```diff
@@ -0,0 +1,623 @@
+    :param args: The namespace object given by argparse.parse()
+    """
+    logger.debug('Trying to auto-determine arguments from environment')
+    if not args.git_sha:
+        commit_id = probe_commit_id()
+        if not commit_id:
+            logger.error('Could not determine commit id. '
+                         'Please set it manually with --git-sha, or ensure you are in a cloned '
+                         'mxnet repository directory')
+            sys.exit(1)
+        args.git_sha = commit_id
+
+    if not args.variant:
+        variant = probe_mxnet_variant(args.libmxnet)
+        if not variant:
+            logger.error('Could not determine mxnet variant. Please set it manually with --variant')
+            sys.exit(1)
+        args.variant = variant
+
+    if not args.os:
+        operating_system = probe_operating_system()
+        if not operating_system:
+            logger.error('Could not determine operating system. Please set it manually with --os')
+            sys.exit(1)
+        args.os = operating_system
+
+    if not args.bucket:
+        artifact_repo_bucket = probe_artifact_repository_bucket()
+        if not artifact_repo_bucket:
+            logger.error('Could not determine artifact repository bucket. Please set it manually with --bucket')
+            sys.exit(1)
+        args.bucket = artifact_repo_bucket
+
+    return args
+
+
+def get_s3_key_prefix(args: argparse.Namespace, subdir: str = '') -> str:
+    """
+    Returns the S3 key prefix given the arguments namespace
+    :param args: The arguments passed in by the user or derived by the script
+    :param subdir: An optional subdirectory in which to store the files. Post-pended to the end of the prefix.
+    :return: A string containing the S3 key prefix to be used to uploading and downloading files to the artifact repository
+    """
+    prefix = "{git_sha}/{lib_type}/{os}/{variant}/".format(**vars(args))
+    if subdir:
+        return "{}{}/".format(prefix, subdir)
+    return prefix
+
+
+def push_artifact(args: argparse.Namespace):
+    """
+    Pushes the artifact to the artifact repository
+    :param args: The arguments passed in to this script by the user
+    :return 0 for success, non-zero for failure
+    """
+
+    args = probe(args)
+
+    logger.info('Pushing artifact with: ')
+    logger.info('COMMIT ID   : {}'.format(args.git_sha))
+    logger.info('OS          : {}'.format(args.os))
+    logger.info('VARIANT     : {}'.format(args.variant))
+    logger.info("LIBMXNET    : {}".format(args.libmxnet))
+    logger.info("LICENSES    : {}".format(args.licenses))
+    logger.info("DEPENDENCIES: {}".format(args.dependencies))
+    logger.info("")
+
+    # Upload mxnet
+    try:
+        logger.info('Uploading libmxnet library...')
+        s3_upload(args.bucket, get_s3_key_prefix(args), [args.libmxnet])
+        logger.info("")
+
+        # Upload licenses, if necessary
+        if args.licenses:
```

Comment:
When would that not be necessary?

Reply:
When the user doesn't define any through the CLI options.

Reply:
We should not make that case an option. We *always* (due to legal reason) have to distribute the licenses as part of an artifact.

Reply:
Sure. I've added it.